sdks/python/apache_beam/runners/dataflow/internal/dependency.py

```diff
@@ -574,23 +631,51 @@ def _download_pypi_sdk_package(temp_dir):
                        'or install a valid {} distribution.'
                        .format(package_name))
-
-  # Get a source distribution for the SDK package from PyPI.
-  cmd_args = [
-      _get_python_executable(), '-m', 'pip', 'download', '--dest', temp_dir,
-      '%s==%s' % (package_name, version),
-      '--no-binary', ':all:', '--no-deps']
-  logging.info('Executing command: %s', cmd_args)
-  processes.check_call(cmd_args)
-  zip_expected = os.path.join(
-      temp_dir, '%s-%s.zip' % (package_name, version))
-  if os.path.exists(zip_expected):
-    return zip_expected
-  tgz_expected = os.path.join(
-      temp_dir, '%s-%s.tar.gz' % (package_name, version))
-  if os.path.exists(tgz_expected):
-    return tgz_expected
-  raise RuntimeError(
-      'Failed to download a source distribution for the running SDK. Expected '
-      'either %s or %s to be found in the download folder.' % (
-          zip_expected, tgz_expected))
+  if fetch_binary:
+    # Get a wheel distribution for the SDK from PyPI.
+    cmd_args = [
+        _get_python_executable(), '-m', 'pip', 'download', '--dest', temp_dir,
+        '%s==%s' % (package_name, version),
+        '--only-binary', ':all:', '--no-deps',
+        '--python-version', language_version_tag,
+        '--implementation', language_implementation_tag, '--abi', abi_tag,
+        '--platform', platform_tag
+    ]
+    logging.info('Executing command: %s', cmd_args)
+    try:
+      processes.check_call(cmd_args)
+    except subprocess.CalledProcessError as e:
+      raise RuntimeError(repr(e))
+    # Example wheel: apache_beam-2.4.0-cp27-cp27mu-manylinux1_x86_64.whl
+    whl_expected = os.path.join(
+        temp_dir,
+        '%s-%s-%s%s-%s-%s.whl' % (package_name.replace('-', '_'), version,
+                                  language_implementation_tag,
+                                  language_version_tag, abi_tag, platform_tag)
+    )
+    if os.path.exists(whl_expected):
+      return whl_expected
+    raise RuntimeError(
+        'Failed to download requested binary distribution for the running SDK. '
+        'Expected %s to be found in the download folder.' % whl_expected)
+  else:
+    # Get a source distribution for the SDK package from PyPI.
+    cmd_args = [
+        _get_python_executable(), '-m', 'pip', 'download', '--dest', temp_dir,
+        '%s==%s' % (package_name, version),
+        '--no-binary', ':all:', '--no-deps']
+
+    logging.info('Executing command: %s', cmd_args)
+    processes.check_call(cmd_args)
+    zip_expected = os.path.join(
+        temp_dir, '%s-%s.zip' % (package_name, version))
+    if os.path.exists(zip_expected):
+      return zip_expected
+    tgz_expected = os.path.join(
+        temp_dir, '%s-%s.tar.gz' % (package_name, version))
+    if os.path.exists(tgz_expected):
+      return tgz_expected
+    raise RuntimeError(
+        'Failed to download a distribution for the running SDK. Expected '
```

Comment:
<!--new_thread; commit:bef56a8915a35b7e1f52ae8e458ea8a589f6e445; resolved:0-->
Should we keep "source distribution", since the other if branch has a corresponding error mentioning "binary distribution"?

Reply:
Thanks, done.